clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -377,36 +389,41 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
         bool have_cond_branch_target = false;
         addr_t cond_branch_target = 0;
         if (shard->prev_instr_.instr.addr != 0 /*first*/ &&
-            type_is_instr_direct_branch(shard->prev_instr_.instr.type) &&
-            // We do not bother to support legacy traces without encodings.
-            TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, shard->file_type_)) {
-            addr_t trace_pc = shard->prev_instr_.instr.addr;
+            type_is_instr_direct_branch(shard->prev_instr_.instr.type) && has_encoding) {
             if (shard->prev_instr_.instr.encoding_is_new)
                 shard->branch_target_cache.erase(trace_pc);
             auto cached = shard->branch_target_cache.find(trace_pc);
             if (cached != shard->branch_target_cache.end()) {
                 have_cond_branch_target = true;
                 cond_branch_target = cached->second;
             } else {
-                instr_t instr;
-                instr_init(GLOBAL_DCONTEXT, &instr);
-                const app_pc decode_pc =
-                    const_cast<app_pc>(shard->prev_instr_.instr.encoding);
-                const app_pc next_pc =
-                    decode_from_copy(GLOBAL_DCONTEXT, decode_pc,
-                                     reinterpret_cast<app_pc>(trace_pc), &instr);
-                if (next_pc == nullptr || !opnd_is_pc(instr_get_target(&instr))) {
+                if ((shard->prev_instr_decoded_ == nullptr ||
+                     !opnd_is_pc(instr_get_target(shard->prev_instr_decoded_)))) {
+
                     // Neither condition should happen but they could on an invalid
                     // encoding from raw2trace or the reader so we report an
                     // invariant rather than asserting.
                     report_if_false(shard, false, "Branch target is not decodeable");
                 } else {
                     have_cond_branch_target = true;
-                    cond_branch_target =
-                        reinterpret_cast<addr_t>(opnd_get_pc(instr_get_target(&instr)));
+                    cond_branch_target = reinterpret_cast<addr_t>(
+                        opnd_get_pc(instr_get_target(shard->prev_instr_decoded_)));
                     shard->branch_target_cache[trace_pc] = cond_branch_target;
                 }
-                instr_free(GLOBAL_DCONTEXT, &instr);
+            }
+        }
+        bool saw_repeated_syscall_instrs_with_same_pc = false;
+        if (shard->prev_instr_.instr.addr != 0 /*first*/) {
+            if (next_pc != nullptr) {
```

Comment:
Would be more readable to check cur_instr_decoded != nullptr here (see other comment about setting cur_instr_decoded to nullptr if decoding fails).